TokenBucket.php

```diff
@@ -0,0 +1,148 @@
+<?
+
+namespace iFixit\TokenBucket;
+
+use \InvalidArgumentException;
+
+require_once __DIR__ . DIRECTORY_SEPARATOR . 'TokenRate.php';
+require_once __DIR__ . DIRECTORY_SEPARATOR . 'Backend.php';
+
+/**
+ * Implements the token bucket algorithm and stores tokens in storage.
+ *
+ * The rate of token regeneration is specified by the TokenRate class. This
+ * also specifies the maximum number of tokens the bucket can hold.
+ *
+ * The current number of tokens is time sensative so it is not stored as a
+ * member of the class, instead you will have to call getTokens() to calculate
+ * and retrieve the current number of tokens.
+ *
+ * Last consume can possibly be changed by another concurrent request, and
+ * therefore is also not stored in the Object's state and always retrieved from
+ * storage when needed.
+ */
+class TokenBucket {
+   // String to identify this bucket uniquely in storage.
+   private $key;
+   // TokenRate is the rate of regeneration for the TokenBucket.
+   private $rate;
+   // Where the token bucket is to be stored
+   private $backend;
+
+   /**
+    * @param $identifier string to uniqely store the bucket.
+    */
+   public function __construct($key, Backend $backend, TokenRate $rate) {
+      if (!is_string($key)) {
+         throw new InvalidArgumentException("identifier must be a string");
+      }
+
+      $this->key = $key;
+      $this->backend = $backend;
+      $this->rate = $rate;
+   }
+
+   /**
+    * Tries to consume a token, if a token can't be consumed, then the interval
+    * of seconds for when the consume will be valid.
+    *
+    * @param amount is an integer amount of tokens that will be attempted to be
+    * consumed
+    *
+    * @return array index 0 being a boolean whether or not a token was consumed
+    *
+    *               index 1 being a double interval of seconds when there
+    *               will be enough tokens to consume the amount requested.
+    */
+   public function consume($amount) {
+      if (!is_numeric($amount) || $amount < 0) {
```

Comment:
The stored bucket seems to support either double or int, and the rest of the operations on tokens support it now too, so I'll update the comment to numeric.